.ci/SetupDevEnv/src/SetupDevEnv.jl

```diff
@@ -1,113 +1,659 @@
 
-    if !isempty(modified_urls)
-        local info_str = "Found following env variables"
-        for (pkg_name, url) in modified_urls
-            info_str *= "\n  " * pkg_name * "=" * url
+# Returns
+
+Returns a list of sets. The index position stands for the round in which the leaf was found,
+e.g. pkg_ordering[1] stands for the first round. The set contains all leafs that were found in the
+round. There is no order within a round.
+"""
+function get_package_dependecy_list(
+    graph::Dict, stop_package::AbstractString=""
+)::Vector{Set{String}}
+    pkg_ordering = _get_package_dependecy_list!(graph, stop_package)
+
+    with_logger(debuglogger) do
+        io = IOBuffer()
+        for i in keys(pkg_ordering)
+            println(io, "$(i): $(pkg_ordering[i])")
         end
-        @info info_str
+        pkg_ordering_str = String(take!(io))
+        @debug "generate dependency ordering list:\n$(pkg_ordering_str)"
     end
 
-    # add all dependencies as develop version to the current julia environment
-    for dep in dependencies
-        if haskey(modified_urls, dep)
-            split_url = split(modified_urls[dep], "#")
-            if length(split_url) > 2
-                error("Ill formed url: $(url)")
-            end
+    return pkg_ordering
+end
+
+function _get_package_dependecy_list!(
+    graph::Dict, stop_package::AbstractString
+)::Vector{Set{String}}
+    @info "calculate the correct sequence for adding QED packages"
+    graph_copy = deepcopy(graph)
+    pkg_ordering = Vector{Set{String}}()
+    while true
+        if isempty(keys(graph_copy["QuantumElectrodynamics"]))
+            push!(pkg_ordering, Set{String}(["QuantumElectrodynamics"]))
+            return pkg_ordering
+        end
+        leafs = Set{String}()
+        _search_leaf!(graph_copy, leafs)
+        if stop_package in leafs
+            push!(pkg_ordering, Set{String}([stop_package]))
+            return pkg_ordering
+        end
+        push!(pkg_ordering, leafs)
+    end
+
+    # unreachable
+    return pkg_ordering
+end
 
-            if length(split_url) == 1
-                @info "Pkg.develop(url=\"" * split_url[1] * "\")"
-                Pkg.add(; url=split_url[1])
-            else
-                @info "Pkg.develop(url=\"" *
-                    split_url[1] *
-                    ";\" rev=\"" *
-                    split_url[2] *
-                    "\")"
-                Pkg.add(; url=split_url[1], rev=split_url[2])
+"""
+    calculate_linear_dependency_ordering(
+        package_dependecy_list::Vector{Set{String}}, 
+        required_dependencies::AbstractVector{String}
+    )::Vector{String}
+
+Computes an ordered list of packages that shows how to add packages to a Julia environment without
+adding a package from the list as an implicit dependency of another package from the list for a
+given package.
+
+Sorry for the difficult documentation. It is much easier to compare the input with the output of
+the function to understand it.
```

Comment:
```suggestion

```

Consider adding an example instead, or just leave it as is, and let the reader decide, if it's difficult ;-)